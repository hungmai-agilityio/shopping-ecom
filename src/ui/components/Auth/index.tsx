import Image from 'next/image';

// Constants
import { inter, popping } from '@/constants';

interface AuthFormProps {
  title: string;
  subtitle: React.ReactNode;
  children: React.ReactNode;
}

const AuthForm = ({ title, subtitle, children }: AuthFormProps) => {
  return (
    <div className="flex my-10">
      <Image
        src="/image-form.webp"
        alt="img-form"
        width={805}
        height={781}
        className="md:block hidden"
      />
      <div className="xl:w-form w-full my-auto xl:mx-36 mx-14">
        <div className={`${popping.className}`}>
          <h1 className={`text-4xl font-medium ${inter.className}`}>{title}</h1>
          <p className="text-base mt-5">{subtitle}</p>
          {children}
        </div>
      </div>
    </div>
  );
};

export default AuthForm;
