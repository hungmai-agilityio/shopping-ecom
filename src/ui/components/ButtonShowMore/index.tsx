'use client';

import { useState, useEffect, memo, useCallback } from 'react';

// Components
import {
  Button,
  CardProductAction,
  SkeletonProductList
} from '@/ui/components';

// Interfaces
import { IProduct, IUser } from '@/interface';

// Libs
import { getProductLimit } from '@/libs';

interface ButtonShowMoreProps {
  queryKey?: string;
  user: IUser;
  startQuery: number;
}

const ButtonShowMore = memo(
  ({ queryKey = '', user, startQuery }: ButtonShowMoreProps) => {
    const [products, setProducts] = useState<IProduct[]>([]);
    const [isLoading, setIsLoading] = useState<boolean>(false);
    const [hasMoreData, setHasMoreData] = useState<boolean>(true);
    const [start, setStart] = useState<number>(startQuery);
    const limit = 4;

    useEffect(() => {
      setProducts([]);
      setHasMoreData(true);
      setStart(startQuery);
    }, [queryKey, startQuery]);

    /**
     *  Handle show more product when click button
     *  Callback when fetching limited data from the last data previously fetched
     **/
    const handleShowMore = useCallback(async () => {
      setIsLoading(true);
      const { data } = await getProductLimit(queryKey, start, limit);

      if (!data.length) {
        setHasMoreData(false);
        setIsLoading(false);
        return;
      }

      setProducts((prev) => [...prev, ...data]);
      setStart((prevStart) => prevStart + limit);
      setIsLoading(false);
    }, [start]);

    return (
      <div className="w-full mt-5">
        <div className="flex flex-wrap lg:gap-x-32 gap-14 lg:justify-normal justify-center">
          {products.map((product) => (
            <CardProductAction
              key={product.id}
              id={product.id}
              name={product.name}
              image={product.image}
              price={product.price}
              ratings={product.ratings}
              colors={product.colors || []}
              isDiscount={product.discount !== undefined}
              discount={product.discount}
              isNew={product.isNew}
              user={user}
              originalPrice={product.originalPrice}
              isNewProduct={product.isNew}
            />
          ))}
          {isLoading && <SkeletonProductList />}
        </div>

        <div className="mt-5 flex justify-center">
          <Button onClick={handleShowMore} disabled={isLoading || !hasMoreData}>
            {isLoading
              ? 'Loading...'
              : hasMoreData
                ? 'Show More'
                : 'No More Data'}
          </Button>
        </div>
      </div>
    );
  }
);

export default ButtonShowMore;
