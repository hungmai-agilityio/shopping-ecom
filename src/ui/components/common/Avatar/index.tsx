import { clsx } from 'clsx';
import Image, { ImageProps } from 'next/image';

// Constants
import { avatarSizeClasses, SIZE } from '@/constants';

interface AvatarProps extends ImageProps {
  size?: SIZE.SMALL | SIZE.MEDIUM | SIZE.LARGE;
  styles?: string;
  circle?: boolean;
  onClick?: () => void;
}

const Avatar = ({
  src,
  alt,
  size = SIZE.SMALL,
  styles,
  circle,
  onClick
}: AvatarProps) => {
  const defaultAvatar = '/user.svg';

  return (
    <div
      className={clsx('relative', avatarSizeClasses[size], styles, {
        'cursor-pointer': onClick
      })}
    >
      <Image
        data-testid="avatar"
        onClick={onClick}
        src={src || defaultAvatar}
        alt={alt}
        fill
        className={clsx(
          'object-cover font-bold',
          circle && 'rounded-full',
          styles
        )}
      />
    </div>
  );
};

export default Avatar;
