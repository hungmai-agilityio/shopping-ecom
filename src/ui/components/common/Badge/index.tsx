interface BadgeProps {
  quantity: number;
  children: React.ReactNode;
}

const Badge = ({ quantity, children }: BadgeProps) => {
  return (
    <div className="relative">
      {quantity > 0 && (
        <span className="absolute -top-3 -right-3 h-6 w-6 rounded-full flex justify-center items-center bg-primary text-white text-[10px] cursor-pointer z-10 overflow-hidden font-semibold">
          {quantity > 99 ? '99+' : quantity}
        </span>
      )}
      {children}
    </div>
  );
};

export default Badge;
