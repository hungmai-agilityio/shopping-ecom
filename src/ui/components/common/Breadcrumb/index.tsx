'use client';

import React from 'react';
import { usePathname } from 'next/navigation';
import { clsx } from 'clsx';
import Link from 'next/link';

interface BreadcrumbProps {
  separator?: string;
  color?: string;
}

const Breadcrumb = ({
  separator = '/',
  color = 'text-gray-700'
}: BreadcrumbProps) => {
  const pathname = usePathname();
  const paths = ['Home', ...pathname.split('/').filter(Boolean)];
  const items = paths.length;

  const displayPaths =
    items > 3 ? [paths[0], '...', paths[items - 2], paths[items - 1]] : paths;

  return (
    <nav className={clsx('flex text-sm', color)}>
      {displayPaths.map((path, index) => (
        <div key={index} className="flex items-center">
          <Link
            href={`/${paths.slice(1, index + 1).join('/')}`}
            className={clsx(
              index === displayPaths.length - 1
                ? 'font-semibold text-black'
                : 'hover:underline'
            )}
          >
            {path}
          </Link>
          {index < displayPaths.length - 1 && (
            <span className="mx-2 text-gray-400">{separator}</span>
          )}
        </div>
      ))}
    </nav>
  );
};

export default Breadcrumb;
