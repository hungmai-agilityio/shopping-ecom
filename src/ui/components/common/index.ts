export { default as Avatar } from '@/ui/components/common/Avatar';
export { default as Button } from '@/ui/components/common/Button';
export { default as Icon } from '@/ui/components/common/Icon';
export { default as Social } from '@/ui/components/common/Icon/Social';
export { default as Card } from '@/ui/components/common/Card';
export { default as CardPromo } from '@/ui/components/common/Card/Promo';
export { default as CardSolution } from '@/ui/components/common/Card/Solution';
export { default as CardProduct } from '@/ui/components/common/Card/Product';
export { default as Tag } from '@/ui/components/common/Tag';
export { default as Badge } from '@/ui/components/common/Badge';
export { default as Breadcrumb } from '@/ui/components/common/Breadcrumb';
export { default as Navbar } from '@/ui/components/common/Navbar';
export { default as Input } from '@/ui/components/common/Input';
export { default as SendEmail } from '@/ui/components/common/Input/SendEmail';
export { default as Logo } from '@/ui/components/common/Logo';
export { default as List } from '@/ui/components/common/List';
export { default as Countdown } from '@/ui/components/common/Countdown';
export { default as Rating } from '@/ui/components/common/Rating';
export { default as SizePicker } from '@/ui/components/common/SizePicker';
export { default as ColorPicker } from '@/ui/components/common/ColorPicker';
export { default as Carousel } from '@/ui/components/common/Carousel';
export { default as Quantity } from '@/ui/components/common/Quantity';
export { default as Dropdown } from '@/ui/components/common/Dropdown';
export { default as Table } from '@/ui/components/common/Table';

export * from '@/ui/components/common/Table/Body';
export * from '@/ui/components/common/Table/Cell';
export * from '@/ui/components/common/Table/Head';
export * from '@/ui/components/common/Table/Row';
export * from '@/ui/components/common/Typo';
