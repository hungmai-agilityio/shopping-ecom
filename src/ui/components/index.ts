export * from '@/ui/components/common';

export { default as Categories } from '@/ui/components/Categories';
