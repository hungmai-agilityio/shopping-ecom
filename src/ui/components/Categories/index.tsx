'use client';
import { clsx } from 'clsx';
import { memo } from 'react';

// Constants
import { categoryLabels, popping } from '@/constants';

// Interfaces
import { ICategory } from '@/interface';

// Components
import { Icon } from '@/ui/components';

interface CategoriesProps {
  categories: ICategory[];
  selected?: string;
  onSelect?: (category: string) => void;
}

const Categories = memo(
  ({ categories, selected, onSelect }: CategoriesProps) => {
    const handleClick = (category: string) => {
      if (onSelect) onSelect(category);
    };

    return (
      <div className={clsx('flex gap-5', popping.className)}>
        {categories.map((category) => (
          <div
            key={category.id}
            onClick={handleClick.bind(null, category.title)}
            className={clsx(
              'cursor-pointer group flex flex-col justify-center items-center w-44 h-36 border border-black rounded-sm transition duration-300',
              {
                'bg-primary text-white border-transparent':
                  selected === category.title,
                'hover:bg-primary hover:border-transparent':
                  selected !== category.title
              }
            )}
          >
            <Icon
              src={category.image}
              alt={category.title}
              width={56}
              height={56}
              className={clsx('transition duration-300', {
                'brightness-0 invert': selected === category.title,
                'group-hover:brightness-0 group-hover:invert':
                  selected !== category.title
              })}
            />
            <span
              className={clsx(
                'mt-5 text-dark transition duration-300 capitalize',
                {
                  'text-white': selected === category.title,
                  'group-hover:text-white': selected !== category.title
                }
              )}
            >
              {categoryLabels[category.title] || category.title}
            </span>
          </div>
        ))}
      </div>
    );
  }
);

export default Categories;
