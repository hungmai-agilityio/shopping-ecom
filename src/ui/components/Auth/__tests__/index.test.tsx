import { TYPE } from '@/constants';
import AuthForm from '@/ui/components/Auth';
import { Input } from '@/ui/components/common';
import { render, screen } from '@testing-library/react';

describe('AuthForm component', () => {
  const propsDefault = {
    title: 'Enter some text',
    subtitle: 'Enter some subtitle',
    children: (
      <div className="flex flex-col gap-5">
        <Input
          label="Text Input"
          placeholder="Enter some text"
          variant={TYPE.SECOND}
        />
      </div>
    )
  };

  test('Should render with default props', () => {
    render(<AuthForm {...propsDefault} />);

    expect(screen.getByText('Enter some text')).toBeInTheDocument();
  });

  test('Should be render match to snapshot', () => {
    const { container } = render(<AuthForm {...propsDefault} />);

    expect(container).toMatchSnapshot();
  });
});
