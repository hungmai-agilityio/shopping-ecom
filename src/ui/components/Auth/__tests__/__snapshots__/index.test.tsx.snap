// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm component Should be render match to snapshot 1`] = `
<div>
  <div
    class="flex my-10"
  >
    <div
      class="lg:w-1/2"
    >
      <img
        alt="img-form"
        class="w-full h-auto object-contain md:block hidden"
        data-nimg="1"
        decoding="async"
        height="600"
        loading="lazy"
        src="/_next/image?url=%2Fimage-form.webp&w=1080&q=75"
        srcset="/_next/image?url=%2Fimage-form.webp&w=640&q=75 1x, /_next/image?url=%2Fimage-form.webp&w=1080&q=75 2x"
        style="color: transparent;"
        width="500"
      />
    </div>
    <div
      class="lg:w-form w-full my-auto lg:mx-36 mx-14"
    >
      <div
        class="className"
      >
        <h1
          class="text-4xl font-medium className"
        >
          Enter some text
        </h1>
        <p
          class="text-base mt-5"
        >
          Enter some subtitle
        </p>
        <div
          class="flex flex-col gap-5"
        >
          <div
            class="flex flex-col gap-2"
          >
            <label
              class="text-sm font-medium text-gray-600 capitalize"
            >
              Text Input
            </label>
            <div
              class="relative"
            >
              <input
                class="w-full p-2 text-gray-700 outline-none border-b border-gray-300 focus:border-dark h-12"
                placeholder="Enter some text"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
